src/mesh/p4est_mesh.jl

```diff
@@ -0,0 +1,273 @@
+                 each dimension if the boundaries in this dimension are periodic.
+"""
+function P4estMesh(trees_per_dimension, coordinates_min, coordinates_max;
+                   polydeg, initial_refinement_level=0, periodicity=true)
+  RealT = promote_type(eltype(coordinates_min), eltype(coordinates_max))
+  mapping = coordinates2mapping(coordinates_min, coordinates_max)
+
+  return P4estMesh(trees_per_dimension, mapping, polydeg=polydeg; RealT=RealT,
+                   initial_refinement_level=initial_refinement_level, periodicity=periodicity)
+end
+
+
+# Create a new p4est_connectivity, which represents a structured rectangle with periodic boundaries.
+# Similar to p4est_connectivity_new_brick, but doesn't use Morton ordering
+# TODO non-periodic
+function connectivity_structured_periodic(cells_x, cells_y)
+  linear_indices = LinearIndices((cells_x, cells_y))
+
+  # Vertices represent the coordinates of the forest. This is used by p4est
+  # to write VTK files.
+  # Trixi doesn't use p4est's coordinates, so the vertices can be empty.
+  num_vertices = 0
+  num_trees = cells_x * cells_y
+  # No corner connectivity is needed
+  num_corners = 0
+  vertices = C_NULL
+  tree_to_vertex = C_NULL
+
+  # Periodic boundaries
+  tree_to_tree = Matrix{Int32}(undef, 4, num_trees)
+  for cell_y in 1:cells_y, cell_x in 1:cells_x
+    tree = linear_indices[cell_x, cell_y]
+    tree_to_tree[1, tree] = linear_indices[mod(cell_x - 2, cells_x) + 1, cell_y] - 1
+    tree_to_tree[2, tree] = linear_indices[mod(cell_x, cells_x) + 1, cell_y] - 1
+    tree_to_tree[3, tree] = linear_indices[cell_x, mod(cell_y - 2, cells_y) + 1] - 1
+    tree_to_tree[4, tree] = linear_indices[cell_x, mod(cell_y, cells_y) + 1] - 1
+  end
+
+  tree_to_face = Matrix{Int8}(undef, 4, num_trees)
+  tree_to_face[1, :] .= 1
+  tree_to_face[2, :] .= 0
+  tree_to_face[3, :] .= 3
+  tree_to_face[4, :] .= 2
+
+  tree_to_corner = C_NULL
+  ctt_offset = Array{Int32}([0])
+
+  corner_to_tree = C_NULL
+  corner_to_corner = C_NULL
+
+  p4est_connectivity_new_copy(num_vertices, num_trees, num_corners,
+                              vertices, tree_to_vertex,
+                              tree_to_tree, tree_to_face,
+                              tree_to_corner, ctt_offset,
+                              corner_to_tree, corner_to_corner)
+end
+
+
+# Check if mesh is periodic
+isperiodic(mesh::P4estMesh) = all(mesh.periodicity)
+isperiodic(mesh::P4estMesh, dimension) = mesh.periodicity[dimension]
+
+@inline Base.ndims(::P4estMesh{NDIMS}) where NDIMS = NDIMS
+@inline Base.real(::P4estMesh{NDIMS, RealT}) where {NDIMS, RealT} = RealT
+
+@inline ntrees(mesh::P4estMesh) = mesh.p4est.trees.elem_count
+@inline ncells(mesh::P4estMesh) = mesh.p4est_mesh.local_num_quadrants
+
+
+function Base.show(io::IO, ::P4estMesh{NDIMS, RealT}) where {NDIMS, RealT}
+  print(io, "P4estMesh{", NDIMS, ", ", RealT, "}")
+end
```

Comment:
```suggestion
function Base.show(io::IO, mesh::P4estMesh)
  print(io, "P4estMesh{", ndims(mesh), ", ", real(mesh), "}")
end
```
Just bikeshedding... but if you have these functions, it would be nice to use them.